ulauncher/modes/extensions/extension_registry.py

```diff
@@ -0,0 +1,69 @@
+from __future__ import annotations
+
+import logging
+from typing import Iterator
+
+from ulauncher.modes.extensions import extension_finder
+from ulauncher.modes.extensions.extension_controller import (
+    ExtensionController,
+    ExtensionNotFoundError,
+    lifecycle_events,
+)
+
+logger = logging.getLogger()
+_ext_controllers: dict[str, ExtensionController] = {}
+
+
+def get(ext_id: str) -> ExtensionController:
+    """Get an extension from the registry or raise if not found."""
+    controller = _ext_controllers.get(ext_id)
+    if not controller:
+        msg = f"Extension with ID '{ext_id}' not found in registry"
+        raise ExtensionNotFoundError(msg)
+
+    return controller
+
+
+def iterate() -> Iterator[ExtensionController]:
+    """Iterate over registered extension controllers."""
+    yield from _ext_controllers.values()
+
+
+def load(ext_id: str, path: str | None = None) -> ExtensionController:
+    """Load an extension controller into the registry without starting it."""
+    if not path:
+        path = extension_finder.locate(ext_id)
+        if not path:
+            _ext_controllers.pop(ext_id, None)
+            msg = f"Extension with ID '{ext_id}' not found"
+            raise ExtensionNotFoundError(msg)
+
+    controller = ExtensionController(ext_id, path)
+    _ext_controllers[ext_id] = controller
+    return controller
+
+
+def load_all() -> Iterator[ExtensionController]:
+    """Load all extensions found by the extension finder into the registry."""
+    for ext_id, ext_path in extension_finder.iterate():
+        load(ext_id, ext_path)
+
+    yield from _ext_controllers.values()
+
+
+@lifecycle_events.on
+def installed(controller: ExtensionController) -> None:
+    """Handle install events dispatched by controllers."""
+    _ext_controllers[controller.id] = controller
+
+
+@lifecycle_events.on
+def updated(controller: ExtensionController) -> None:
+    """Handle update events dispatched by controllers."""
+    _ext_controllers[controller.id] = controller
```

Comment:
As I mentioned elsewhere I don't think we need `updated` as lifecycle event, because `_ext_controllers[controller.id]` will already be `controller`

Maybe with your upcoming PR for the preview feature this will be needed in that one though? (I would still suggest removing it from this PR and adding it there in that case).